### ISSUE TITLE
OAS 3.1 support?

### ISSUE THREAD

**Issue by Kormyr** [42Crunch/vscode-openapi#110]
Hello,

I just discovered this wonderful VSCode extension. Thank you for all the work on this.

OpenAPI Specification 3.1 has been released recently (https://github.com/OAI/OpenAPI-Specification/releases/tag/3.1.0).
Are there already plans to implement it?

**Comment by ak1394** [comment]
Thanks! Happy you liked the extension. As for 3.1, we plan to support it of course. My guess it will not make it to the next release (which quite close to being published) but one after that.

**Comment by ak1394** [comment]
I had a closer look at the new spec, and it looks like there are some issues in around schema validation which might delay it a bit. 

The JSON schema for 3.1 is WIP https://github.com/OAI/OpenAPI-Specification/pull/2016 but once it's done it'll require 2019-09 schema support in VS Code https://github.com/microsoft/vscode/issues/98724

**Comment by spacedsweden** [comment]
https://github.com/microsoft/vscode/issues/98724#issuecomment-786502628

Can we do yaml meanwhile?

**Comment by ak1394** [comment]
Nope, VS Code uses the same JSON Schema implementation for both YAML and JSON. You're linking to a comment mentioning some other non vs-code implementations of JSON Schema 2020-12, but that doesn't help much as we need VS Code one (which provides auto-completion, etc) to implement the new schema standard.

**Comment by goodoldneon** [comment]
@ak1394 can you go into a little more detail on your blocker(s) for supporting 3.1.0? Your extension has been a fantastic guardrail for our new-to-OpenAPI engineers, but we have to start using 3.1.0 (to get proper nullable reference schemas: https://github.com/OAI/OpenAPI-Specification/issues/1368)

**Comment by ak1394** [comment]
@goodoldneon thanks for the feedback! We would like to introduce 3.1 support as soon as we can, the main issue being  2019-09 schema support in VS Code...

**Comment by apkrieg** [comment]
Came looking for 3.1.0 support, found sadness. Can't wait for an update.

**Comment by spacedsweden** [comment]
Any news on this? is the spec stable enough now?

**Comment by spacedsweden** [comment]
I have done the same, i am actually using their vs extension as well. however i do miss the sidepanel i get with 42crunch

**Comment by PhilParisot** [comment]
Any news? 3.1.0 has been out for a while now.

**Comment by ak1394** [comment]
Well, there seem to be some progress on implementing required schema support: https://github.com/microsoft/vscode/issues/98724

As soon as it lands in VS Code, we'll start working on it.

**Comment by rubensa** [comment]
Have [this](https://github.com/microsoft/vscode/pull/150001/files) fixed the impediment to get OAS 3.1 support?

**Comment by ak1394** [comment]
Nope, the issue to watch is the one I've linked above, it is still open.

**Comment by anikitin** [comment]
This plugin can use both Swagger UI and ReDoc to preview the spec. Swagger UI still doesn't support OAS 3.1 but ReDoc does.
Would be nice to have at least 3.1 support with only ReDoc implementation.

**Comment by defunctzombie** [comment]
Is there a way to patch the plugin so it will work with 3.1 spec files? Even if not all features are present?

**Comment by behrenle** [comment]
Any news on this topic?

**Comment by ak1394** [comment]
No news, there is not much happening in https://github.com/microsoft/vscode/issues/98724 so we have to wait as well

**Comment by kkamath-bandwidth** [comment]
FYI. I was happily using the 42Crunch plugin, but this ended up being the forcing function to move to next option in VSCode marketplace. The ReDocly one forces you to use an apiKey for a simple preview, no thanks.

**Comment by spacedsweden** [comment]
What’s the problem using an api key?

**Comment by kkamath-bandwidth** [comment]
I did not want to go through the step of configuring it. So personal preference, and there is no problem.

**Comment by anywhereinfo** [comment]
We have a custom spectral file and redocly does not seem to support it.

**Comment by spacedsweden** [comment]
no you would need to port it to redoc linter

**Comment by cuttingclyde** [comment]
Swagger UI open source project released support for OAS 3.1 on 6-12-23, https://github.com/swagger-api/swagger-ui.

I'm not sure why you would also need to wait for VSCode support for JSON Schema 2020-12 validation in OAS 3.1, to simply support swagger.io preview of 3.1 versioned OAS documents. Can't they live separately? Only OAS 3.0 validation with OAS 3.1 preview?

**Comment by anywhereinfo** [comment]
For me, its important to have a editor with intellisense and schema validation for 3.1.  Creation of OAS 3.1 document and Governance via Spectral Linter can be done without swagger preview. But if a tool can support all  the use cases why not. So, i found stoplight studio, which is opensource, [suppo](https://github.com/stoplightio/studio), supports 3.1 editing, intellisense and preview. Plus it leverages spectral linting, so my existing investment in rules can be ported without any change

**Comment by shreyasbharath** [comment]
Bump, any news on this please?

**Comment by kornicameister** [comment]
https://github.com/42Crunch/vscode-openapi/pull/242 got merged recently.
Does it mean we can get new tag in the air?

**Comment by ak1394** [comment]
I'll look at adding support for 3.1 for preview (which should be enabled by new SwaggerUI) but until https://github.com/microsoft/vscode/issues/98724 is done, we won't have proper 3.1 support in the editor...

**Comment by ak1394** [comment]
No, unfortunately there has been no progress on the vscode issue since May 2022. It could be that we'll have to do something about 3.1 support despite not having working schema support in VS Code, but I have no concrete time frame for it.

**Comment by sbeitzel** [comment]
So, as @terjew mentions above, it looks as though MS has decided to export all the JSON stuff to a separate language service [microsoft/vscode-json-languageservice](https://github.com/microsoft/vscode-json-languageservice). The discussion on the perpetually [linked issue](https://github.com/microsoft/vscode/issues/98724) seems to amount to various end users coming in with variants of, "Hey, this plugin I like doesn't work with modern JSON and it's because VS Code doesn't know how to use it? Can I haz fix?" and with the responsible developer saying, "There's no bug, here, and there's a language service so maybe the plugin or someone can go ask it for stuff?"

Anyway. I have no idea what a language service does (other than crash and require a restart when I put a typo in my source code) nor how easily a plugin like this one could consume such a service, nor even if that hypothetical interaction would solve our problems. But does anyone else reading this issue know?

**Comment by andreasvh-conceto** [comment]
Hi @sbeitzel 
i am subscribed to this issue and still interested. Last week i had to create an OAS 3.1. I was forced to use the nextSwaggerEditor. JsonSchemaV4 is not enough. Still hoping for support from vscode, so that plugins like this will work with the newest standard.

**Comment by alexanderilyin** [comment]
THIMC: Solution for me was to move to https://huma.rocks/

**Comment by sraesch** [comment]
Is it a joke? I mean no progress at all for 4 years in support OAS 3.1?! Sorry, but this turns the tool useless for me and I had to uninstall it.

**Comment by ViliusZur** [comment]
Any news? Swaggerhub is horrible for creating big Open API specifications. Maybe someone car recommend some other tool that supports OAS 3.1 until this one is updated?

**Comment by anthony-42crunch** [comment]
Hi,
I can confirm support for OAS 3.1 in this VSCode extension is under development. 
I expect it will be released soon and will share an update once a release date is confirmed by the dev team.

**Comment by anywhereinfo** [comment]
I have been using combination of spectral and redocly for 2 years now to
manage OAS 3.1. Both are free

-- Maninder Batth
No trees were killed in the creation of this message. However, many
electrons were terrible inconvenienced.

**Comment by anthony-42crunch** [comment]
Quick preview of the OAS 3.1 support. [watch](https://developers.42crunch.com/t/oas-3-1-preview/150)
Should be available in the OpenAPI Editor VSCode plugin soon. 
Will keep you posted.

**Comment by visini** [comment]
I saw that #316 was merged. Very exciting. Is there any way to try this, or should we wait for the release of a new version? Thank you!

**Comment by ak1394** [comment]
@visini we plan to release the new version early next week, so it shouldn't be a very long wait. Other than that, you can build it from source yourself, but it's a bit more work...

**Comment by anthony-42crunch** [comment]
OpenAPI (Swagger) Editor v4.32 is now available which includes a beta release for OAS 3.1 support.  

<img width="476" alt="Image" src="https://github.com/user-attachments/assets/e984f004-fe38-4ea7-aa99-0a290317382e" />

**Comment by ak1394** [comment]
@anywhereinfo You're welcome to share your OpenAPI file with the issues you've encountered. The VS Code unfortunately still doesn't support 2019-09 Schema dialect, so we had to rewrite official OpenAPI 3.1 schema in the older schema dialect supported by VS Code, and it's quite possible that there are bugs in it.

**Comment by abraxxa** [comment]
For me the seemingly only issue is that it doesn't recognize $ref as valid.

**Comment by ak1394** [comment]
@abraxxa Could you provide an example? The is one currently known issue where it doesn't recognize $ref in the top level schemas in the components (ie. when a top level schema immediately references another one) but it would be great to confirm if you are experiencing the same issue.

**Comment by abraxxa** [comment]
Mainly $ref used in the response object.
I can give an exact example in the evening.

**Comment by ak1394** [comment]
@abraxxa That would be very helpful!

**Comment by anywhereinfo** [comment]
It also does not handle examples, when used in conjunction with ref

-- Maninder Batth
No trees were killed in the creation of this message. However, many
electrons were terrible inconvenienced.

**Comment by ak1394** [comment]
@anywhereinfo Thanks!

**Comment by abraxxa** [comment]
The first $ref `"$ref": "#/components/schemas/User"` has the problem 'Property $ref is not allowed.', the other two are fine.

```
"responses": {
          "201": {
            "description": "Created",
            "content": {
              "application/json": {
                "schema": {
                  "type": "object",
                  "additionalProperties": false,
                  "properties": {
                    "data": {
                      "$ref": "#/components/schemas/User"
                    }
                  }
                }
              }
            }
          },
          "400": {
            "description": "Bad Request",
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/ErrorResponse"
                }
              }
            }
          },
          "403": {
            "$ref": "#/components/responses/Error"
          }
        },
```

**Comment by ak1394** [comment]
@abraxxa @anywhereinfo Thanks for your feedback guys, I've published version 4.32.1 that should fix the issues you've reported. However, please let me know if I've missed anything.

**Comment by abraxxa** [comment]
I can confirm that the warnings are gone. Thanks for the quick fix!

**Comment by ak1394** [comment]
@abraxxa thanks a lot for checking!
